In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

In [2]:
df = pd.read_csv(r"F:\zomato.csv")
df.shape

(51717, 17)

## OUTCOME: 
There are 51717 rows and 17 columns in the DataSet

In [3]:

df.head()

url  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             address                   name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  online_order book_table   rate  votes                             phone  \
0          Yes        Yes  4.1/5    775    080 42297555\r\n+91 9743772233   
1          Yes         No  4.1/5    787                      080 41714161   
2          Yes         No  3.8/5    918                    +91 9663487993   
3           No         No  3.7/5     88                    +91 9620009302   
4           No         No  3.8/5    166  +91 8026612447\r\n+91 9901210005   

       location            rest_type  \
0  Banashankari        Casual Dining   
1  Banashankari        Casual Dining   
2  Banashankari  Cafe, Casual Dining   
3  Banashankari          Quick Bites   
4  Basavanagudi        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines approx_cost(for two people)  \
0  North Indian, Mughlai, Chinese                         800   
1     Chinese, North Indian, Thai                         800   
2          Cafe, Mexican, Italian                         800   
3      South Indian, North Indian                         300   
4        North Indian, Rajasthani                         600   

                                        reviews_list menu_item  \
0  [('Rated 4.0', 'RATED\n  A beautiful place to ...        []   
1  [('Rated 4.0', 'RATED\n  Had been here for din...        []   
2  [('Rated 3.0', "RATED\n  Ambience is not that ...        []   
3  [('Rated 4.0', "RATED\n  Great food and proper...        []   
4  [('Rated 4.0', 'RATED\n  Very good restaurant ...        []   

  listed_in(type) listed_in(city)  
0          Buffet    Banashankari  
1          Buffet    Banashankari  
2          Buffet    Banashankari  
3          Buffet    Banashankari  
4          Buffet    Banashankari

# Checking the columns in the dataset

In [4]:
df.columns.to_list()

['url',
 'address',
 'name',
 'online_order',
 'book_table',
 'rate',
 'votes',
 'phone',
 'location',
 'rest_type',
 'dish_liked',
 'cuisines',
 'approx_cost(for two people)',
 'reviews_list',
 'menu_item',
 'listed_in(type)',
 'listed_in(city)']

## Columns description:

     -url:contains the url of the resturant in the zomato dataset
     -address:contains the address of the resturant in Bengaluru
     -name:contains name of the resturant 
     -online_order:wheather online order is available in the resturant or not
     -book_table: book option available or not
     -rate:contains overall rating of the resturant out of 5
     -votes:contains total number of rating for the resturant
     -phone:contains the phone number of the resturant
     -location:contains the neighbourhood in which the resturant is located
     -rest_type:type of the resturant
     -dish_liked:dishes people liked in the resturant
     -cuisines:food styles,sepreted by comma
     -approx_cost(for two people):contains the approximate cost for meal for two people
     -reviews_list:list of tuples containing reviews for the resturant 
     -menu_item:contains list of menus available in the returant
     -listed_in(type):type of meal
     -listed_in(city):contains the neighbourhood in which the resturant is listed

In [5]:
df.rename(columns={'approx_cost(for two people)':'approx_cost', 'listed_in(type)':'Type'}, inplace = True)

In [6]:
df.rename(columns={'listed_in(city)':'area'},inplace=True)

In [7]:
df.columns

Index(['url', 'address', 'name', 'online_order', 'book_table', 'rate', 'votes',
       'phone', 'location', 'rest_type', 'dish_liked', 'cuisines',
       'approx_cost', 'reviews_list', 'menu_item', 'Type', 'area'],
      dtype='object')

In [8]:
df.drop(['url','address', 'menu_item','reviews_list','phone'],axis=1, inplace = True)

In [9]:
df.head()

name online_order book_table   rate  votes      location  \
0                  Jalsa          Yes        Yes  4.1/5    775  Banashankari   
1         Spice Elephant          Yes         No  4.1/5    787  Banashankari   
2        San Churro Cafe          Yes         No  3.8/5    918  Banashankari   
3  Addhuri Udupi Bhojana           No         No  3.7/5     88  Banashankari   
4          Grand Village           No         No  3.8/5    166  Basavanagudi   

             rest_type                                         dish_liked  \
0        Casual Dining  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1        Casual Dining  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Cafe, Casual Dining  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3          Quick Bites                                        Masala Dosa   
4        Casual Dining                                Panipuri, Gol Gappe   

                         cuisines approx_cost    Type          area  
0  North Indian, Mughlai, Chinese         800  Buffet  Banashankari  
1     Chinese, North Indian, Thai         800  Buffet  Banashankari  
2          Cafe, Mexican, Italian         800  Buffet  Banashankari  
3      South Indian, North Indian         300  Buffet  Banashankari  
4        North Indian, Rajasthani         600  Buffet  Banashankari

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   name          51717 non-null  object
 1   online_order  51717 non-null  object
 2   book_table    51717 non-null  object
 3   rate          43942 non-null  object
 4   votes         51717 non-null  int64 
 5   location      51696 non-null  object
 6   rest_type     51490 non-null  object
 7   dish_liked    23639 non-null  object
 8   cuisines      51672 non-null  object
 9   approx_cost   51371 non-null  object
 10  Type          51717 non-null  object
 11  area          51717 non-null  object
dtypes: int64(1), object(11)
memory usage: 4.7+ MB


In [11]:
df['rate'].unique()

array(['4.1/5', '3.8/5', '3.7/5', '3.6/5', '4.6/5', '4.0/5', '4.2/5',
       '3.9/5', '3.1/5', '3.0/5', '3.2/5', '3.3/5', '2.8/5', '4.4/5',
       '4.3/5', 'NEW', '2.9/5', '3.5/5', nan, '2.6/5', '3.8 /5', '3.4/5',
       '4.5/5', '2.5/5', '2.7/5', '4.7/5', '2.4/5', '2.2/5', '2.3/5',
       '3.4 /5', '-', '3.6 /5', '4.8/5', '3.9 /5', '4.2 /5', '4.0 /5',
       '4.1 /5', '3.7 /5', '3.1 /5', '2.9 /5', '3.3 /5', '2.8 /5',
       '3.5 /5', '2.7 /5', '2.5 /5', '3.2 /5', '2.6 /5', '4.5 /5',
       '4.3 /5', '4.4 /5', '4.9/5', '2.1/5', '2.0/5', '1.8/5', '4.6 /5',
       '4.9 /5', '3.0 /5', '4.8 /5', '2.3 /5', '4.7 /5', '2.4 /5',
       '2.1 /5', '2.2 /5', '2.0 /5', '1.8 /5'], dtype=object)

In [12]:
#removing "NEW" , "_" and "/5" from rate column
def handlerate(value):
    if(value=="NEW" or value=='-'):
        return np.nan
    else:
        value=str(value).split('/')
        value=value[0]
        return float(value)
    
df['rate']=df['rate'].apply(handlerate)
df['rate'].head()
        

0    4.1
1    4.1
2    3.8
3    3.7
4    3.8
Name: rate, dtype: float64

In [13]:
df['approx_cost'].unique()

array(['800', '300', '600', '700', '550', '500', '450', '650', '400',
       '900', '200', '750', '150', '850', '100', '1,200', '350', '250',
       '950', '1,000', '1,500', '1,300', '199', '80', '1,100', '160',
       '1,600', '230', '130', '50', '190', '1,700', nan, '1,400', '180',
       '1,350', '2,200', '2,000', '1,800', '1,900', '330', '2,500',
       '2,100', '3,000', '2,800', '3,400', '40', '1,250', '3,500',
       '4,000', '2,400', '2,600', '120', '1,450', '469', '70', '3,200',
       '60', '560', '240', '360', '6,000', '1,050', '2,300', '4,100',
       '5,000', '3,700', '1,650', '2,700', '4,500', '140'], dtype=object)

In [14]:
def handlecomma(value):
    value = str(value)
    if ',' in value:
        value = value.replace(',', '')
        return float(value)
    else:
        return float(value)
    
df['approx_cost'] = df['approx_cost'].apply(handlecomma)
df['approx_cost'].unique()

array([ 800.,  300.,  600.,  700.,  550.,  500.,  450.,  650.,  400.,
        900.,  200.,  750.,  150.,  850.,  100., 1200.,  350.,  250.,
        950., 1000., 1500., 1300.,  199.,   80., 1100.,  160., 1600.,
        230.,  130.,   50.,  190., 1700.,   nan, 1400.,  180., 1350.,
       2200., 2000., 1800., 1900.,  330., 2500., 2100., 3000., 2800.,
       3400.,   40., 1250., 3500., 4000., 2400., 2600.,  120., 1450.,
        469.,   70., 3200.,   60.,  560.,  240.,  360., 6000., 1050.,
       2300., 4100., 5000., 3700., 1650., 2700., 4500.,  140.])

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   name          51717 non-null  object 
 1   online_order  51717 non-null  object 
 2   book_table    51717 non-null  object 
 3   rate          41665 non-null  float64
 4   votes         51717 non-null  int64  
 5   location      51696 non-null  object 
 6   rest_type     51490 non-null  object 
 7   dish_liked    23639 non-null  object 
 8   cuisines      51672 non-null  object 
 9   approx_cost   51371 non-null  float64
 10  Type          51717 non-null  object 
 11  area          51717 non-null  object 
dtypes: float64(2), int64(1), object(9)
memory usage: 4.7+ MB


In [16]:
df.isnull().sum()

name                0
online_order        0
book_table          0
rate            10052
votes               0
location           21
rest_type         227
dish_liked      28078
cuisines           45
approx_cost       346
Type                0
area                0
dtype: int64

In [17]:
df=df[df['rest_type'].notna()]
df=df[df['location'].notna()]
df=df[df['cuisines'].notna()]
df=df[df['approx_cost'].notna()]
df.isnull().sum()

name                0
online_order        0
book_table          0
rate             9885
votes               0
location            0
rest_type           0
dish_liked      27713
cuisines            0
approx_cost         0
Type                0
area                0
dtype: int64

In [18]:
df['rate'].describe()

count    41263.000000
mean         3.702011
std          0.439985
min          1.800000
25%          3.400000
50%          3.700000
75%          4.000000
max          4.900000
Name: rate, dtype: float64

In [19]:
q1=df['rate'].quantile(0.25)
q3=df['rate'].quantile(0.75)

iqr = q3 - q1

lower_range = iqr = q1-(1.5 * iqr)
upper_range = iqr = q3+(1.5 * iqr)

outliers = len(df[(df['rate']<lower_range)|(df['rate']>upper_range)])
print("Number of Rows having outliers: ", outliers)

Number of Rows having outliers:  183


In [20]:
df['rate'].fillna(df['rate'].median(), inplace=True)
df.isnull().sum()

name                0
online_order        0
book_table          0
rate                0
votes               0
location            0
rest_type           0
dish_liked      27713
cuisines            0
approx_cost         0
Type                0
area                0
dtype: int64

In [21]:
'''df.reset_index(inplace = True)
df.drop(['index'],axis = 1, inplace = True)'''

"df.reset_index(inplace = True)\ndf.drop(['index'],axis = 1, inplace = True)"

In [22]:
df['online_order'] = df['online_order'].replace({'Yes':1,"No":0})
df.head()

name  online_order book_table  rate  votes      location  \
0                  Jalsa             1        Yes   4.1    775  Banashankari   
1         Spice Elephant             1         No   4.1    787  Banashankari   
2        San Churro Cafe             1         No   3.8    918  Banashankari   
3  Addhuri Udupi Bhojana             0         No   3.7     88  Banashankari   
4          Grand Village             0         No   3.8    166  Basavanagudi   

             rest_type                                         dish_liked  \
0        Casual Dining  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1        Casual Dining  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Cafe, Casual Dining  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3          Quick Bites                                        Masala Dosa   
4        Casual Dining                                Panipuri, Gol Gappe   

                         cuisines  approx_cost    Type          area  
0  North Indian, Mughlai, Chinese        800.0  Buffet  Banashankari  
1     Chinese, North Indian, Thai        800.0  Buffet  Banashankari  
2          Cafe, Mexican, Italian        800.0  Buffet  Banashankari  
3      South Indian, North Indian        300.0  Buffet  Banashankari  
4        North Indian, Rajasthani        600.0  Buffet  Banashankari

In [23]:
df['book_table'] = df['book_table'].replace({'Yes':1,"No":0})
df.head()

name  online_order  book_table  rate  votes      location  \
0                  Jalsa             1           1   4.1    775  Banashankari   
1         Spice Elephant             1           0   4.1    787  Banashankari   
2        San Churro Cafe             1           0   3.8    918  Banashankari   
3  Addhuri Udupi Bhojana             0           0   3.7     88  Banashankari   
4          Grand Village             0           0   3.8    166  Basavanagudi   

             rest_type                                         dish_liked  \
0        Casual Dining  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1        Casual Dining  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Cafe, Casual Dining  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3          Quick Bites                                        Masala Dosa   
4        Casual Dining                                Panipuri, Gol Gappe   

                         cuisines  approx_cost    Type          area  
0  North Indian, Mughlai, Chinese        800.0  Buffet  Banashankari  
1     Chinese, North Indian, Thai        800.0  Buffet  Banashankari  
2          Cafe, Mexican, Italian        800.0  Buffet  Banashankari  
3      South Indian, North Indian        300.0  Buffet  Banashankari  
4        North Indian, Rajasthani        600.0  Buffet  Banashankari

In [24]:
df.drop(columns=['name','area'],axis=1,inplace = True)

In [25]:
df.head()

online_order  book_table  rate  votes      location            rest_type  \
0             1           1   4.1    775  Banashankari        Casual Dining   
1             1           0   4.1    787  Banashankari        Casual Dining   
2             1           0   3.8    918  Banashankari  Cafe, Casual Dining   
3             0           0   3.7     88  Banashankari          Quick Bites   
4             0           0   3.8    166  Basavanagudi        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines  approx_cost    Type  
0  North Indian, Mughlai, Chinese        800.0  Buffet  
1     Chinese, North Indian, Thai        800.0  Buffet  
2          Cafe, Mexican, Italian        800.0  Buffet  
3      South Indian, North Indian        300.0  Buffet  
4        North Indian, Rajasthani        600.0  Buffet

In [26]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, LabelEncoder

In [27]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

In [28]:
x= df.drop('rate',axis = 1)
y=df['rate']

In [29]:
x

online_order  book_table  votes                    location  \
0                 1           1    775                Banashankari   
1                 1           0    787                Banashankari   
2                 1           0    918                Banashankari   
3                 0           0     88                Banashankari   
4                 0           0    166                Basavanagudi   
...             ...         ...    ...                         ...   
51712             0           0     27                  Whitefield   
51713             0           0      0                  Whitefield   
51714             0           0      0                  Whitefield   
51715             0           1    236  ITPL Main Road, Whitefield   
51716             0           0     13  ITPL Main Road, Whitefield   

                 rest_type                                         dish_liked  \
0            Casual Dining  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1            Casual Dining  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2      Cafe, Casual Dining  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3              Quick Bites                                        Masala Dosa   
4            Casual Dining                                Panipuri, Gol Gappe   
...                    ...                                                ...   
51712                  Bar                                                NaN   
51713                  Bar                                                NaN   
51714                  Bar                                                NaN   
51715                  Bar                       Cocktails, Pizza, Buttermilk   
51716   Bar, Casual Dining                                                NaN   

                                     cuisines  approx_cost           Type  
0              North Indian, Mughlai, Chinese        800.0         Buffet  
1                 Chinese, North Indian, Thai        800.0         Buffet  
2                      Cafe, Mexican, Italian        800.0         Buffet  
3                  South Indian, North Indian        300.0         Buffet  
4                    North Indian, Rajasthani        600.0         Buffet  
...                                       ...          ...            ...  
51712                             Continental       1500.0  Pubs and bars  
51713                             Finger Food        600.0  Pubs and bars  
51714                             Finger Food       2000.0  Pubs and bars  
51715                             Finger Food       2500.0  Pubs and bars  
51716  Finger Food, North Indian, Continental       1500.0  Pubs and bars  

[51148 rows x 9 columns]

In [30]:
y.head()

0    4.1
1    4.1
2    3.8
3    3.7
4    3.8
Name: rate, dtype: float64

In [31]:
one_hot_cols = ['cuisines', 'rest_type']
le_cols = ['location','Type','dish_liked']

In [43]:
x = preprocessor.fit_transform(x)

In [44]:
x.shape

(51148, 8143)

In [32]:
ct = ColumnTransformer([
        ('one_hot', OneHotEncoder(), one_hot_cols),
        ('label', LabelEncoder(), le_cols)
    ], remainder='passthrough')

In [33]:
ct.fit_transform(x)

TypeError: fit_transform() takes 2 positional arguments but 3 were given

In [34]:
# Define the columns to be one-hot encoded
one_hot_cols = ['cuisines', 'rest_type']

# Define the columns to be label encoded
le_cols = ['location', 'Type', 'dish_liked']

# Define the column transformer
ct = ColumnTransformer([
    ('one_hot', OneHotEncoder(), one_hot_cols),
    ('label', LabelEncoder(), le_cols)
], remainder='passthrough')

# Fit and transform the data
transformed_data = ct.fit_transform(x)


TypeError: fit_transform() takes 2 positional arguments but 3 were given

In [35]:
x.head()

online_order  book_table  votes      location            rest_type  \
0             1           1    775  Banashankari        Casual Dining   
1             1           0    787  Banashankari        Casual Dining   
2             1           0    918  Banashankari  Cafe, Casual Dining   
3             0           0     88  Banashankari          Quick Bites   
4             0           0    166  Basavanagudi        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines  approx_cost    Type  
0  North Indian, Mughlai, Chinese        800.0  Buffet  
1     Chinese, North Indian, Thai        800.0  Buffet  
2          Cafe, Mexican, Italian        800.0  Buffet  
3      South Indian, North Indian        300.0  Buffet  
4        North Indian, Rajasthani        600.0  Buffet

In [36]:
y.head()

0    4.1
1    4.1
2    3.8
3    3.7
4    3.8
Name: rate, dtype: float64

In [ ]:
lr_model=LinearRegression()
lr_model.fit(x_train,y_train)

In [ ]:
from sklearn.metrics import r2_score
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
y_predict=lr_model.predict(x_test)
r2_score(y_test,y_predict)

In [ ]:
from sklearn.ensemble import RandomForestRegressor
RF_Model=RandomForestRegressor(n_estimators=650,random_state=245,min_samples_leaf=.0001)
RF_Model.fit(x_train,y_train)
y_predict=RF_Model.predict(x_test)
r2_score(y_test,y_predict)

In [ ]:
#Preparing Extra Tree Regression
from sklearn.ensemble import  ExtraTreesRegressor
ET_Model=ExtraTreesRegressor(n_estimators = 1000)
ET_Model.fit(x_train,y_train)
y_predict=ET_Model.predict(x_test)
r2_score(y_test,y_predict)